### PR TITLE
Configure timeouts for the distributor

### PR DESCRIPTION
**Pull request by dogukanteber**
This PR is just the beginning of the task titled "Ensure proper timeouts are configured for each service". At this point, it only implements timeouts for the distributor.

I did not want to advance further to check this is the correct way to implement proper timeouts

### CODE REVIEW

File: gateway/gateway.go
Comment (dogukanteber) [inline gateway/gateway.go:47]: Since this is just to make sure I am doing the right thing, I passed empty values to this function. I will fix it as soon as I get an approval

**Review by friedrichg — COMMENTED** [review]
Looks good. I think

File: gateway/proxy.go
Comment (dogukanteber) [inline gateway/proxy.go:20]: I gave default values as 5's but it was random. What should be the correct timeout values for the distributor?

Comment (friedrichg) [reply gateway/proxy.go:20]: Only testing can tell. It's fine for now

**Comment by dogukanteber** [comment]
This PR is now complete. I implemented the timeouts for the distributor and query-frontend. I will create another PR for the ruler and alertmanager when #6 is merged

File: gateway/proxy.go
Comment (friedrichg) [inline gateway/proxy.go:80]:
```suggestion
	t := http.DefaultTransport.(*http.Transport).Clone()
	t.ResponseHeaderTimeout = readTimeout
	// add more timeouts
	return t
```
Transport != DefaultTransport. see https://github.com/golang/go/blob/go1.20.3/src/net/http/transport.go#L38-L54

**Review by friedrichg — CHANGES_REQUESTED** [review]
Ideally, we should not repeat structures. Use **Upstream** everywhere.
Not redefine them every time. I made a few suggestions, but there are more.

File: gateway/config.go
Comment (dogukanteber) [reply gateway/config.go:41]: I was going to do that in a separate PR. I did not want this PR to grow more but if handling this issue in this PR is better, that is fine with me

File: gateway/proxy.go
Comment (dogukanteber) [reply gateway/proxy.go:80]: Good to know but I wonder if using a global variable is a good idea. It may lead to bugs that are hard to trace.

If we are going to use the same config as `DefaultTransport`, I propose creating our own `Transport` struct inside `NewProxy` function. What do you think?

Comment (friedrichg) [reply gateway/proxy.go:80]: what do you mean? t is not a global variable and it's calling `Clone()` or am I missing something?

File: gateway/config.go
Comment (friedrichg) [reply gateway/config.go:41]: It's too much duplication, there is no value. It's easier if there is less code to review, even for you. 
Let's stay on this PR to simplify this.

File: gateway/proxy.go
Comment (dogukanteber) [reply gateway/proxy.go:80]: Oh, no I missed something 😅. I focused too much on `DefaultTransport` and missed `Clone()`.

File: gateway/config.go
Comment (dogukanteber) [reply gateway/config.go:41]: Okay 👍

**Comment by dogukanteber** [comment]
I messed up this branch. So, I opened another PR in a separate branch #18. Closing this one.

**Pull request closed without merging**
